ManalHamdi/assignment4/Assignment4.cpp

```diff
@@ -0,0 +1,79 @@
+#include <iostream>
+#include <utility>
+#include <queue>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+bool isValidNeighbour(const std::vector<std::vector<bool> > &map_tiles, int row, int column, int number_rows, int number_columns) {
+	return (map_tiles[row][column] && row >= 0 && row < number_rows && column >= 0 && column < number_columns);
+}
+std::vector<std::pair<int, int> > getNeighboursIndices(const std::vector<std::vector<bool> > &map_tiles, int row, int column, int number_rows, int number_columns) {
+	std::vector<std::pair<int, int> > neighbours;
+	
+	int delta_rows[] = {1, -1, 0, 0};
+	int delta_cols[] = {0, 0, 1, -1};
+	
+	for (int i = 0; i < 4; i++) {
+		if (isValidNeighbour(row + delta_rows[i], column + delta_cols[i], number_rows, number_columns)) {
+			neighbours.push_back(std::make_pair(row + delta_rows[i], column + delta_cols[i]));
+		}
+	}
+	return neighbours;
+}
+void BFS(const std::vector<std::vector<bool> > &map_tiles,  int start_row,  int start_column, int number_rows,  int number_columns, std::vector<std::vector<bool> > &visited) {
+	std::queue <std::pair<int, int> > q;
+	q.push(std::make_pair(start_row,start_column));
+	while (!q.empty()) {
+		std::pair<int, int> current = q.front();
+		q.pop();
+		if (visited[current.first][current.second]) {
+			continue;
+		}
+		visited[current.first][current.second] = true;
+
+		std::vector<std::pair<int,int> > neighbours = getNeighboursIndices(map_tiles, current.first, current.second, number_rows, number_columns);
+		for (std::pair<int, int> neighbour: neighbours) {
+			q.push(neighbour);
+		}
+	}
+}
+int countNumberIslands(std::vector<std::vector<bool> > &map_tiles, int number_rows, int number_columns) {
```

Comment:
Since your code is pretty much perfect, how about doing some error handling?

For example, in this assignment you assume that the `number_rows` and `number_columns` you're given will match nicely to the dimensions of the map. When you're writing 'real world' code, it might be harmful to make assumptions like these. Someone somewhere *will* try to pass an empty map with random dimensions or something similar and then your code will explode, or what's worse, return something weird that won't be noticed straight away until it's months later and nobody is quite sure what could be wrong anymore.

When I interview people for jobs at Google, this is one of the things I look out for, how prepared is the person do deal with unexpected input or whether they just make assumptions without much consideration for their implications. So when you interview make sure to always point out everything that can go wrong and write defensive code that is able to handle bad input to some degree.

In this case, you can just run over the map and calculate the dimensions yourself without needing the two ints.
With ints in general, if you're sure the value can't be negative, declare them as `unsigned int`, won't help with all edge cases, but at least it's something :)

Reply:
Also, I don't think you're modifying the map tiles anywhere, can you make it `const` to make it clear?